pkg/runs/runsDownload.go

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright contributors to the Galasa project
+ */
+package runs
+
+import (
+	"bufio"
+	"context"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/galasa.dev/cli/pkg/api"
+	galasaErrors "github.com/galasa.dev/cli/pkg/errors"
+	"github.com/galasa.dev/cli/pkg/galasaapi"
+	"github.com/galasa.dev/cli/pkg/utils"
+)
+
+// DownloadArtifacts - performs all the logic to implement the `galasactl runs download` command,
+// but in a unit-testable manner.
+func DownloadArtifacts(
+	runName string,
+	forceDownload bool,
+	fileSystem utils.FileSystem,
+	timeService utils.TimeService,
+	console utils.Console,
+	apiServerUrl string) error {
+
+	var err error = nil
+	var runs []galasaapi.Run
+	var artifactPaths []string
+
+	runs, err = GetRunsFromRestApi(runName, timeService, apiServerUrl)
+	if err == nil {
+		for _, run := range runs {
+			runId := run.GetRunId()
+			artifactPaths, err = GetArtifactPathsFromRestApi(runId, apiServerUrl)
+			if err == nil {
+				for _, artifactPath := range artifactPaths {
+					artifactData, err := GetFileFromRestApi(runId, strings.TrimPrefix(artifactPath, "/"), apiServerUrl)
+					if err != nil {
+						return err
+					} else {
+						err = WriteArtifactToFileSystem(fileSystem, runName, artifactPath, artifactData, forceDownload)
+						if err != nil {
+							return err
+						}
+					}
+				}
+			}
+		}
+	}
+	return err
+}
+
+// Retrieves the paths of all artifacts for a given test run using its runId.
+func GetArtifactPathsFromRestApi(runId string, apiServerUrl string) ([]string, error) {
+
+	var err error = nil
+	var artifactPaths []string
+	log.Println("Retrieving artifact paths for the given run")
+
+	// An HTTP client which can communicate with the api server in an ecosystem.
+	restClient := api.InitialiseAPI(apiServerUrl)
+
+	var artifactsList []galasaapi.ArtifactIndexEntry
+	artifactsList, httpResponse, err := restClient.ResultArchiveStoreAPIApi.
```

Comment:
Can any of these methods return nil ? like when they have errors ?
Worried that we'll get a nil de-reference when using `fluent` style of calls.
Best to separate out and check for errors for each piece ?

Reply:
Only the `Execute()` method can return nil - `InitialiseAPI()` and `GetRasRunArtifactByPath` always return objects (an `APIClient` and an `ApiGetRasRunArtifactByPathRequest` object, respectively).